### DISCUSSION TITLE
Uncaught Error: Class "Relay\Relay" not found

### DISCUSSION THREAD

**Discussion by jordantrizz** [rhubarbgroup/.github#14] [Support (ROC)] [answered]
## Description

Installed Relay, added the following to wp-config.php

```
define('WP_REDIS_CLIENT','relay')
```

Everything seemed to be working until running wp-cli

```
Fatal error: Uncaught Error: Class "Relay\Relay" not found in /home/user/sites/website.com/htdocs/wp-content/object-cache.php:775
Stack trace:
#0 /home/user/sites/website.com/htdocs/wp-content/object-cache.php(543): WP_Object_Cache->connect_using_relay()
rhubarbgroup/redis-cache#1 /home/user/sites/website.com/htdocs/wp-content/object-cache.php(256): WP_Object_Cache->__construct()
rhubarbgroup/redis-cache#2 /home/user/sites/website.com/htdocs/wp-includes/load.php(750): wp_cache_init()
rhubarbgroup/redis-cache#3 /home/user/sites/website.com/htdocs/wp-settings.php(131): wp_start_object_cache()
rhubarbgroup/redis-cache#4 phar:///usr/local/bin/wp/vendor/wp-cli/wp-cli/php/WP_CLI/Runner.php(1336): require('...')
rhubarbgroup/redis-cache#5 phar:///usr/local/bin/wp/vendor/wp-cli/wp-cli/php/WP_CLI/Runner.php(1254): WP_CLI\Runner->load_wordpress()
rhubarbgroup/redis-cache#6 phar:///usr/local/bin/wp/vendor/wp-cli/wp-cli/php/WP_CLI/Bootstrap/LaunchRunner.php(28): WP_CLI\Runner->start()
rhubarbgroup/redis-cache#7 phar:///usr/local/bin/wp/vendor/wp-cli/wp-cli/php/bootstrap.php(78): WP_CLI\Bootstrap\LaunchRunner->process()
rhubarbgroup/redis-cache#8 phar:///usr/local/bin/wp/vendor/wp-cli/wp-cli/php/wp-cli.php(32): WP_CLI\bootstrap()
rhubarbgroup/redis-cache#9 phar:///usr/local/bin/wp/php/boot-phar.php(11): include('...')
rhubarbgroup/redis-cache#10 /usr/local/bin/wp(4): include('...')
rhubarbgroup/redis-cache#11 {main}
  thrown in /home/user/sites/website.com/htdocs/wp-content/object-cache.php on line 775
Error: There has been a critical error on this website.Learn more about troubleshooting WordPress. There has been a critical error on this website.
```

## Expected Behavior

There shouldn't be an error.

## Actual Behavior

There's an error.

## Steps to Reproduce

1. Install relay
2. Added define('WP_REDIS_CLIENT','relay') to wp-config.php
3. run wp-cli ex (wp user)

## Environment

* Plugin version:  2.4.1
* PHP version: 7.4
* WordPress version:  6.2.2

**Comment by tillkruss** [comment, marked as answer]
Your PHP version doesn't have Relay installed (or not loaded), otherwise the class would exist.

Follow the installations instructions: https://relay.so/docs/installation

**Comment by jordantrizz** [comment]
If anyone else stumbles upon this, php-relay was installed for php-fpm for PHP 7.4. The default php-cli version was set to PHP 8.0, which is what wp-cli was using, hence the error.

@tillkruss would using try and then catch be a safer way of handling this? Versus a fatal error? Or does that not make sense at all in those case?

**Reply by tillkruss** [reply]
After catching the error, what would you suggest doing/showing?

**Reply by jordantrizz** [reply]
Not proceeding to load the library, falling back to phpredis and generating an error in error_log

**Reply by tillkruss** [reply]
Few people check those, I rather be explicit and crash & burn if someone expects Relay to be used, but it's not present.

**Reply by jordantrizz** [reply]
Totally get it; what about crash and burn with an appropriate error message? Someone who needs help understanding PHP and namespaces would be confused. This crash and burn also disables wp-cli completely, --skip-plugins don't skip object cache drop-ins. What happens if phpredis isn't installed? Does it still crash and burn?

If wp-cli is not functional, lots of hosting providers who use SSO using the wp-cli login-command package would fail, even though the site functions properly. So on services such as GridPane, and SpinupWP where you can select PHP versions from 7.4, 8.0, 8.1 for a given site. The wp-cli PHP version is not site-specific and might not include php-relay or php-redis. 

You're fixing a problem that isn't caused by your plugin, it's a host configuration issue. But you could save people time.

Since using Redis object cache isn't a requirement for WordPress to function, having a crash in burn in-place for php-relay or phpredis not being available means a site will not function versus simply falling back to a disabled state. A warning in the status of the object cache plugin would also help.

What's worse, a downed site or not having redis object cache? I could be overthinking it, but thought I'd give my explanation. Thanks for all you do Till!

**Reply by tillkruss** [reply]
Yeah, catching it with a more better exception about how to ensure Relay is running would be good. Do you want to open a PR about it?
